Initiate scAdvanced cheatsheet

sjspielman opened the pull request:
Towards #646 

This PR begins the process of establishing a scAdvanced cheatsheet. I am going to split this up into 2 PRs, since there's a decent amount to review here.
This PR accomplishes two goals:
- Add functions to the intro cheatsheet that should have been in the intro cheatsheet all along
- Establish the scAdvanced cheatsheet, including all tables of functions

There will be one additional PR to add in `Seurat <-> SCE` conversions (stubbed out but not filled out in the cheatsheet!). Right now, I have only committed `md` files since I expect changes! At the end of this PR, I'd like for the _intro PDF_ to make it into the repo, but the advanced PDF will be part of the next PR.

Some specific notes for reviewers:
- Enjoy link clicking and finding all the links I didn't get right 🥴! That's it for this bullet.
- Turns out `map_df()` is in the intro r cheatsheet, but it's only used in _setup code_ for the intro workshop. It didn't seem like a big deal to me to leave in. I had forgotten there was some very light `purrr` in intro, too!!
- Any overall thoughts on order of the topics in the cheatsheet? I tried to loosely structure in order of when items appear (though this isn't possible to do fully linearly), with all `tidyverse` at the end. The conversion is stubbed out at the top as this is a common question.
- `plotExpression()` is really the only "new"  `scater` function in scAdvanced. But, I went ahead and added some of intro the `scater` functions into the advanced, since they're important for advanced and come up a lot (`plotReducedDim` and `plotUMAP`). Eventually this will get solidified with "cheatsheets by topic," so I thought this duplication would be ok for now. 
- I couldn't quite land on the right level of detail for `miQC` explanations, any thoughts? Somehow "Plot the miQC model" for `plotModel()` didn't seem enough, but actually explaining what gets plotted makes for a LONG table row...!
- I'm not a fan, but the title for `HarmonyMatrix()` _is_ "Main Harmony interface". Feeling like I just ignore that and write "harmony matrix"...
- Currently all `tidyverse` is in one table, and I both feel like that's reasonable and it's excessive! Therefore, how many tables this is can be up to reviewers! (Maybe a separate table for just `ggplot2`?)

jashapiro commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -0,0 +1,115 @@
+
+<br>
+
+| Library/Package| Piece of Code| What it's called| What it does |
```
For organization: I think I would separate out `dplyr/tidyr/tibble` from `purrr` from `ggplot2`

sjspielman replied:
Done in https://github.com/AlexsLemonade/training-modules/pull/663/commits/024f202677e51798fa9bd81c73d6f478a52c15ad

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -39,19 +39,29 @@ Read the `SingleCellExperiment` package documentation (and e-book) [**here**](ht
 | `SingleCellExperiment` | [`SingleCellExperiment()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html) | Single Cell Experiment| Creates a `SingleCellExperiment` object  |
 | `SingleCellExperiment`| [`colData()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances) | Column Data | Extracts and stores cell-level metadata that describes features of the `SingleCellExperiment` object|
 | `SingleCellExperiment`| [`rowData()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances)  | Row Data   | Extracts and stores gene-level metadata that describes features of the `SingleCellExperiment` object|
+| `SingleCellExperiment`| [`assays()` and `assay()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances)  | Assay(s) | Extracts or assigns a given assay(s) from a `SingleCellExperiment` object|
```
Do we want to link to the `SummarizedExperiment`  documentation to explain this and assayNames?(with some kind of note that the same functions apply at the top)? https://rdrr.io/bioc/SummarizedExperiment/man/SummarizedExperiment-class.html

sjspielman replied:
> (with some kind of note that the same functions apply at the top)

Can you clarify ^?

jashapiro replied:
What I meant here was that we are linking to files defined with respect to `SummarizedExperiment` objects, and it may not be immediately clear that they also work on `SingleCellExperiment` objects. So I wondered if we might want a note that says "Some of the functions below are defined in `SummarizedExperiment`, but can be applied to `SingleCellExperiment because the latter is a special version of the former" (Not those words!)

sjspielman replied:
Ah you know, I didn't even notice that! Will do

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -39,19 +39,29 @@ Read the `SingleCellExperiment` package documentation (and e-book) [**here**](ht
 | `SingleCellExperiment`| [`colData()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances) | Column Data | Extracts and stores cell-level metadata that describes features of the `SingleCellExperiment` object|
 | `SingleCellExperiment`| [`rowData()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances)  | Row Data   | Extracts and stores gene-level metadata that describes features of the `SingleCellExperiment` object|
+| `SingleCellExperiment`| [`assays()` and `assay()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances)  | Assay(s) | Extracts or assigns a given assay(s) from a `SingleCellExperiment` object|
+| `SingleCellExperiment`| [`assayNames()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations)  | Assay names  | Returns a vector of the names of all assays in a `SingleCellExperiment` object|
```
While `#3_Adding_low-dimensional_representations` is the first place in that doc that `assayNames()`  appears, I'm not sure it is a good place to link for the docs...

sjspielman replied:
Agreed...Done in https://github.com/AlexsLemonade/training-modules/pull/663/commits/b585b5fb43da26ecc0cbfdd64b24672b91151aba

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -39,19 +39,29 @@ Read the `SingleCellExperiment` package documentation (and e-book) [**here**](ht
 | `SingleCellExperiment`| [`rowData()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances)  | Row Data   | Extracts and stores gene-level metadata that describes features of the `SingleCellExperiment` object|
+| `SingleCellExperiment`| [`assays()` and `assay()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#2_Creating_SingleCellExperiment_instances)  | Assay(s) | Extracts or assigns a given assay(s) from a `SingleCellExperiment` object|
+| `SingleCellExperiment`| [`assayNames()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations)  | Assay names  | Returns a vector of the names of all assays in a `SingleCellExperiment` object|
 | `SingleCellExperiment`| [`logcounts()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations)| Log counts| Stores or extracts log-transformed single-cell experiment count data as an assay of the `SingleCellExperiment` object|
```
Not something you changed, but fixing this link.
```suggestion
| `SingleCellExperiment`| [`logcounts()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#4_Convenient_access_to_named_assays)| Log counts| Stores or extracts log-transformed single-cell experiment count data as an assay of the `SingleCellExperiment` object|
```

jashapiro reviewed (commented):
This looks good to start! 

I think I addressed most of your bullet points with line comments, but for some of the more overall comments, I think the order is fine, though I might move the seurat conversion to the end. But I can probably go either way on that.

I checked many links, but probably not all of them...

sjspielman commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -0,0 +1,115 @@
+|----------------------|----------------------------|--------------------------------------------|--------------------------------------------------------------|
+| `batchelor`| [`MultiBatchPCA()`](https://rdrr.io/bioc/batchelor/man/multiBatchPCA.html)| Multi-batch PCA | Perform PCA across multiple gene expression matrices while accounting for batch differences |
+| `batchelor`| [`fastMNN()`](https://rdrr.io/bioc/batchelor/man/fastMNN.html)| Fast mutual nearest neighbors correction | Perform integration with mutual nearest neighbors using the `fastMNN` algorithm |
+| `harmony`| [`HarmonyMatrix()`](https://rdrr.io/github/immunogenomics/harmony/man/HarmonyMatrix.html)| Main Harmony interface | Perform integration with `harmony` |
```
What's extremely annoying is that this wraps an internal function `RunHarmony`, so I can't really say "run harmony." Going with "perform integration with harmony on a matrix" I suppose.

sjspielman replied:
Done in https://github.com/AlexsLemonade/training-modules/pull/663/commits/c5c13436f0f17a4f6966f6736bfc150dd9034b7e

sjspielman commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -39,19 +39,29 @@ Read the `SingleCellExperiment` package documentation (and e-book) [**here**](ht
+| `SingleCellExperiment`| [`assayNames()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations)  | Assay names  | Returns a vector of the names of all assays in a `SingleCellExperiment` object|
 | `SingleCellExperiment`| [`logcounts()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations)| Log counts| Stores or extracts log-transformed single-cell experiment count data as an assay of the `SingleCellExperiment` object|
 | `SingleCellExperiment`| [`counts()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#4_Convenient_access_to_named_assays)| Counts| Stores or extracts raw single-cell experiment count data as an assay of the `SingleCellExperiment` object|
+| `SingleCellExperiment`| [`reducedDims()` and `reducedDim()`](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#3_Adding_low-dimensional_representations)| Reduced dim(s)| Extracts or assigns a given reduced dimension(s) from a `SingleCellExperiment` object|
```
Done in https://github.com/AlexsLemonade/training-modules/pull/663/commits/1d6ecd72d710cbdff8fa52b2351a7e882bc3a6f1

sjspielman commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -0,0 +1,115 @@
+
+### `scater`
+
+Read the `scater` package documentation [**here**](https://rdrr.io/bioc/scater/), and a vignette on its usage [**here**](http://www.bioconductor.org/packages/release/bioc/vignettes/scater/inst/doc/overview.html).
```
Done across a few commits!

sjspielman commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -0,0 +1,115 @@
+#### The tables below consist of valuable functions or commands that will help you through this module.
+##### Each table represents a different library/tool and its corresponding commands.
+
+##### You may also be interested in the [Introduction to R and Tidyverse cheatsheet](intro-to-R-tidyverse-cheatsheet.pdf) and the [Introduction to Single-Cell RNA sequencing cheatsheet](scRNA-seq-cheatsheet.pdf).
```
Done in https://github.com/AlexsLemonade/training-modules/pull/663/commits/7e05abd6282e6571625982214f60ccda8703fd21

sjspielman commented:
@jashapiro this should be ready for another look! I've implemented the in-line reviews, and I also went ahead and purged all the `here` links from the 3 cheatsheets (intro r, intro sc, advanced sc) since it was moving brackets around. The only TBD is this comment which I wasn't quite sure what you meant? https://github.com/AlexsLemonade/training-modules/pull/663#discussion_r1124764389

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -24,12 +24,16 @@ Read the Base `R` documentation [**here**](https://rdrr.io/r/).
 | Base `R`| [`<-function(x) { <code> }`](https://adv-r.hadley.nz/functions.html) | Function | Creates a function that would take the defined parameters as input and execute the commands within the curly braces  |
```
I realized that we don't have a link to the `\()` anywhere... I don't know if this is the right place, but we should put it somewhere... Maybe with `purrr`? Docs here: https://rdrr.io/r/base/function.html

jashapiro reviewed (commented):
Looks good! But since we are apparently moving the links, I made them bigger and more specific. 

I also realized we don't have syntax `\(x){}` anywhere, which we should. Somewhere. I don't care too much about where, but made a suggestion.

Also, maybe TOCs?

sjspielman commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -0,0 +1,133 @@
+#### The tables below consist of valuable functions or commands that will help you through this module.
+##### Each table represents a different library/tool and its corresponding commands.
+
+##### You may also be interested in the following additional cheatsheets:[Introduction to R and Tidyverse cheatsheet](https://github.com/AlexsLemonade/training-modules/blob/master/module-cheatsheets/intro-to-R-tidyverse-cheatsheet.pdf) and the [Introduction to Single-Cell RNA sequencing cheatsheet](https://github.com/AlexsLemonade/training-modules/blob/master/module-cheatsheets/scRNA-seq-cheatshe ...
```
I accepted this change, but might revert. I'm remembering now the reason I chose _not_ to do the `raw/` links is because it will forcibly download the cheatsheet, rather than linking to an in-browser view (I wonder if this is browser-specific, but I get same behavior in safari+firefox).

I generally like to avoid surprise downloads, so I might update text to say plainly that these are download links, or include a link to download vs view.

sjspielman commented:
This has been updated in a couple ways:

- I added cheatsheet links to the advanced cheatsheet, but not to the intro, under the assumption that we'll do a cheatsheet overall organizing by topic in the not distant future and can update all around at that time
- For all PDF links, I added language that they are downloads
- Added TOCs via `doctoc` and copy/pasting
- Added a code chunk showing anonymous function syntax, feedback appreciated there! The example I chose is inspired by things we do in the integration instruction notebook

I also decided to think a little bit about PDF export, whose current [status quo](https://github.com/AlexsLemonade/training-modules/blob/master/CONTRIBUTING.md#cheatsheets) (please enjoy the `blame` view here and the most recent commit 🥴) is not my favorite - a little too manual, and `MacDown` is no longer being maintained if this [issue](https://github.com/MacDownApp/macdown/issues/1104) is any indication. I ended up settling on the `markdown-pdf` VS Code extension, with some css to override mediocre font colors in `module-cheatsheets/assets/markdown-pdf.css`. I quite like the look of these cheatsheets, and they are at least more reproducible than what we've been doing! If reviewers are on board, I'll file a subsequent PR to update instructions for PDF export. For fun, reviewers may also appreciate this commit https://github.com/AlexsLemonade/training-modules/pull/663/commits/3ff00fd6e3fb438cdf77a9fdb7c836e201e6f7d8 and infer the silly journey I had along the way 😄

sjspielman commented:
> then converts that to pdf via pandoc or some other scriptable tool.

I'll just note that if this involves a custom LaTeX template, that's all you.... I will revert to Macdown-exported PDFs, but will state for the record I find those versions _way_ less attractive than what `markdown-pdf` gives.

jashapiro commented:
> > then converts that to pdf via pandoc or some other scriptable tool.
> 
> I'll just note that if this involves a custom LaTeX template, that's all you.... I will revert to Macdown-exported PDFs, but will state for the record I find those versions _way_ less attractive than what `markdown-pdf` gives.

No latex! I was thinking whatever is being used under the hood here, maybe with a conda environment to install it. 

I'm fine with leaving these as they are here. I just don't want to get into encoding & documenting a new method that may be similarly brittle.

jashapiro commented on module-cheatsheets/scRNA-seq-advanced-cheatsheet.md:
```diff
@@ -0,0 +1,172 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+- [`scater`](#scater)
```
Rendering robustness
```suggestion

- [`scater`](#scater)
```

jashapiro commented on module-cheatsheets/scRNA-seq-cheatsheet.md:
```diff
@@ -9,11 +9,22 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+- [Base `R`](#base-r)
```
Rendering robustness
```suggestion

- [Base `R`](#base-r)
```

jashapiro reviewed (commented):
A couple more suggestions for rendering

jashapiro commented:
> @jashapiro I had missed the TOC and PDF re-render for the `intro` cheatsheet. Can I get a thumbs here for changes made in [f103eb5](https://github.com/AlexsLemonade/training-modules/commit/f103eb5b03bd1cb59001683abdb984b527fde66e) and [4795e88](https://github.com/AlexsLemonade/training-modules/commit/4795e88f26f649944ae7e114cfd7471a3b4f5c91) ?

👍🏼
